File: apps/ios/Data/Sources/Data/SupabaseConnector.swift
import Foundation

// SupabaseConnector — bridges PowerSync to Supabase on iOS.
//
// Ported from packages/db/src/connector.ts (P2.2b).
// Mirrors apps/android/data/.../sync/SupabaseConnector.kt (P2.2a).
//
// - fetchCredentials: gives PowerSync the current Supabase JWT.
// - uploadData: flushes the local write queue to Postgres, preserving order.
//
// Financial-safety note: writes are applied in queue order and retried on
// failure; because the ledger is append-only and server-authoritative,
// replays are safe (no in-place money mutation).
//
// Coalescing: consecutive ops sharing the same table + op type are batched
// into a single PostgREST request (array upsert / id IN (...) delete). This
// turns a bulk import of N transactions into a handful of requests while
// preserving queue ordering (FK ordering like account-before-txn untouched).
// PATCH always stays per-row: two updates to the same table can touch
// different columns and must not be merged.

import PowerSync
import Supabase

/// Postgres schema that holds all PocketCare tables.
public let DB_SCHEMA = "pocketcare"

// MARK: - SyncDiagnostic

/// Optional sink for structured sync failures.
///
/// A module-level hook rather than an init parameter, so this package stays
/// free of any dependency on the app's diagnostics layer. The app opts in by
/// calling setSyncDiagnosticSink at startup.
public struct SyncDiagnostic: Sendable {
    public let table: String
    public let op: String
    public let rows: Int
    public let code: String?
    public let message: String?
    public let cls: String?
    public let attempts: Int?
    public let quarantined: Bool?
}

// Swift 6: nonisolated(unsafe) for a simple nullable callback.
nonisolated(unsafe) private var diagnosticSink: (@Sendable (SyncDiagnostic) -> Void)? = nil

public func setSyncDiagnosticSink(_ fn: (@Sendable (SyncDiagnostic) -> Void)?) {
    diagnosticSink = fn
}

// MARK: - FaultInjection

/// FAULT INJECTION — dev only.
///
/// Makes unreachable paths reachable (e.g. an RLS denial you can't produce
/// on demand). Matches connector.ts's fault injection semantics exactly.
public struct FaultInjection: Sendable {
    /// Unqualified table name, or "*" for everything.
    public let table: String
    /// SQLSTATE to report, e.g. "42501" (RLS) or "23503" (foreign key).
    public let code: String
    public let message: String?
    public let status: Int?

    public init(table: String, code: String, message: String? = nil, status: Int? = nil) {
        self.table = table
        self.code = code
        self.message = message
        self.status = status
    }
}

nonisolated(unsafe) private var fault: FaultInjection? = nil

public func setFaultInjection(_ f: FaultInjection?) { fault = f }
public func getFaultInjection() -> FaultInjection? { fault }

private func injectedErrorFor(table: String) -> FailureDetails? {
    guard let f = fault else { return nil }
    let bare = table.contains(".") ? String(table.split(separator: ".").last ?? Substring(table)) : table
    guard f.table == "*" || f.table == bare else { return nil }
    return FailureDetails(
        code: f.code,
        message: f.message ?? "injected fault (\(f.code)) for \(bare)",
        status: f.status
    )
}

// MARK: - SupabaseConnector

/// Keys bumped this session (in-memory; cleared on success).
nonisolated(unsafe) private var bumped = Set<String>()

private func reportSyncDiagnostic(_ d: SyncDiagnostic) {
    // Diagnostics must never make a sync failure worse than it already is.
    diagnosticSink?(d)
}

/// Convert a PowerSync CrudEntry to a CrudOpSummary for quarantine tracking.
private func toOpSummary(_ entry: CrudEntry) -> CrudOpSummary {
    CrudOpSummary(
        // entry.clientId is a non-optional Int64 in the real SDK (confirmed via
        // PowerSync Swift's DocC index and CrudEntry.swift source: `let clientId:
        // Int64`, no `?`) — unlike Kotlin's nullable clientId, so there's nothing
        // to .map over. Widens implicitly to CrudOpSummary.clientId's Int64? field.
        clientId: entry.clientId,
        table: entry.table,
        op: entry.op.rawValue,
        id: entry.id,
        // opDataTyped, not opData: CrudEntry.opData (confirmed via the real
        // CrudEntry.swift source) is a *deprecated, stringified* accessor —
        // `[String: String?]?` — that turns every value, including money minor
        // units, into a string for backwards compatibility. opDataTyped
        // (`[String: JsonValue]?`) is the real typed data. Using opData here
        // would put stringified numbers into the dead-letter record.
        payloadJson: encodePayload(entry.opDataTyped?.mapValues { anyValue(from: $0) })
    )
}

/// Extract FailureDetails from any thrown error.
private func extractFailure(_ error: Error) -> FailureDetails {
    // supabase-swift wraps PostgREST errors; we extract what we can.
    let msg = error.localizedDescription
    // PostgREST errors often include "code" in their userInfo or message.
    // Without a concrete type match, surface what's available.
    return FailureDetails(code: nil, message: msg, status: nil)
}

public final class SupabaseConnector: PowerSyncBackendConnectorProtocol, @unchecked Sendable {
    private let client: SupabaseClient
    private let powerSyncUrl: String
    private let schema: String

    public init(
        client: SupabaseClient,
        powerSyncUrl: String,
        schema: String = DB_SCHEMA
    ) {
        self.client = client
        self.powerSyncUrl = powerSyncUrl
        self.schema = schema
    }

    public func fetchCredentials() async throws -> PowerSyncCredentials? {
        guard let session = try? await client.auth.session else { return nil }
        return PowerSyncCredentials(
            endpoint: powerSyncUrl,
            token: session.accessToken
        )
    }

    public func uploadData(database: PowerSyncDatabaseProtocol) async throws {
        guard let batch = try await database.getCrudBatch() else { return }

        let ops = batch.crud
        var i = 0

        while i < ops.count {
            let op = ops[i]
            let rel = client.schema(schema).from(op.table)

            // Extend the run while the next op targets the same table + op type.
            // PATCH stays per-row (different columns may be touched).
            var j = i + 1
            while j < ops.count
                && ops[j].table == op.table
                && ops[j].op == op.op
                && op.op != .patch
            {
                j += 1
            }
            let run = Array(ops[i..<j])

            // Fault injection short-circuits before any real network call.
            let injected = injectedErrorFor(table: "\(schema).\(op.table)")
            var failure: FailureDetails? = injected

            if failure == nil {
                do {
                    switch op.op {
                    case .put:
                        // opDataTyped (not opData) — see the comment on toOpSummary:
                        // opData stringifies every value, which would upload money
                        // minor-unit columns as JSON strings instead of integers.
                        var rows: [[String: AnyJSON]] = run.map { entry in
                            var row: [String: AnyJSON] = [:]
                            if let d = entry.opDataTyped {
                                for (k, v) in d { row[k] = anyJSON(from: v) }
                            }
                            row["id"] = .string(entry.id)
                            return row
                        }
                        try await rel.upsert(rows).execute()
                    case .delete:
                        let ids = run.map { $0.id }
                        try await rel.delete().in("id", value: ids).execute()
                    case .patch:
                        var data: [String: AnyJSON] = [:]
                        if let d = op.opDataTyped {
                            for (k, v) in d { data[k] = anyJSON(from: v) }
                        }
                        try await rel.update(data).eq("id", value: op.id).execute()
                    @unknown default:
                        break
                    }
                } catch {
                    failure = extractFailure(error)
                }
            }

            if let failure {
                let summaries = run.map { toOpSummary($0) }
                let verdict = await handleUploadFailure(
                    db: database,
                    run: summaries,
                    failure: failure,
                    onBump: { k in bumped.insert(k) }
                )

                print("[PocketCare sync] upload failed for \(schema).\(op.table) "
                    + "(\(op.op.rawValue), \(run.count) row(s), attempt \(verdict.attempts), "
                    + "\(verdict.classification.cls))"
                    + (verdict.quarantined ? " — moved to Problems syncing" : ""))

                reportSyncDiagnostic(SyncDiagnostic(
                    table: "\(schema).\(op.table)",
                    op: op.op.rawValue,
                    rows: run.count,
                    code: failure.code,
                    message: failure.message,
                    cls: verdict.classification.cls,
                    attempts: verdict.attempts,
                    quarantined: verdict.quarantined
                ))

                if verdict.quarantined {
                    i = j
                    continue
                }
                throw NSError(domain: "PocketCareSync", code: 1,
                              userInfo: [NSLocalizedDescriptionKey: failure.message ?? "Upload failed"])
            }

            // Succeeded — clear retry counters for this run.
            for entry in run {
                let key = opKey(table: entry.table, op: entry.op.rawValue, rowId: entry.id)
                if bumped.remove(key) != nil {
                    await clearAttempts(db: database, key: key)
                }
            }
            i = j
        }

        try await batch.complete()
    }
}

// MARK: - JsonValue → AnyJSON / Any conversion
//
// PowerSync's CrudEntry exposes typed op data as `opDataTyped: JsonParam?`
// (JsonParam = [String: JsonValue]), confirmed via the real source at the
// pinned commit (Sources/PowerSync/Protocol/db/JsonParam.swift). JsonValue
// has its own `toValue() -> Any?`, but that method is NOT `public` in the
// SDK — it's internal to the PowerSync module and inaccessible here — so we
// need our own converters. Two are needed because they target different
// destination types: AnyJSON for supabase-swift's PostgREST builder, and
// plain Any? for Quarantine.swift's encodePayload (JSONSerialization-based).

/// Convert a PowerSync JsonValue into supabase-swift's AnyJSON, preserving the
/// real type (Int/Double/Bool/String/null/array/object) instead of stringifying.
///
/// supabase-swift's real AnyJSON cases (confirmed via source at the pinned
/// commit, Sources/Helpers/AnyJSON/AnyJSON.swift): .null, .bool(Bool),
/// .integer(Int), .double(Double), .string(String), .object(JSONObject),
/// .array(JSONArray). There is no .number or .boolean case — those were
/// guesses, not the real API.
private func anyJSON(from value: JsonValue) -> AnyJSON {
    switch value {
    case .null: return .null
    case .string(let s): return .string(s)
    case .int(let n): return .integer(n)
    case .double(let d): return .double(d)
    case .bool(let b): return .bool(b)
    case .array(let arr): return .array(arr.map { anyJSON(from: $0) })
    case .object(let obj): return .object(obj.mapValues { anyJSON(from: $0) })
    }
}

/// Convert a PowerSync JsonValue into a plain Swift value, for encodePayload's
/// `[String: Any?]?` input (the quarantine dead-letter record).
private func anyValue(from value: JsonValue) -> Any? {
    switch value {
    case .null: return nil
    case .string(let s): return s
    case .int(let n): return n
    case .double(let d): return d
    case .bool(let b): return b
    case .array(let arr): return arr.map { anyValue(from: $0) as Any }
    case .object(let obj): return obj.mapValues { anyValue(from: $0) as Any }
    }
}

File: apps/android/data/src/main/kotlin/com/sanvya/app/data/sync/SupabaseConnector.kt
package com.sanvya.app.data.sync

/**
 * SupabaseConnector — bridges PowerSync to Supabase on Android.
 *
 * Ported from packages/db/src/connector.ts (P2.2a).
 *
 * - fetchCredentials: gives PowerSync the current Supabase JWT.
 * - uploadData: flushes the local write queue to Postgres, preserving order.
 *
 * Financial-safety note: writes are applied in queue order and retried on
 * failure; because the ledger is append-only and server-authoritative, replays
 * are safe (no in-place money mutation).
 *
 * Coalescing: consecutive ops sharing the same table + op type are batched
 * into a single PostgREST request (array upsert / id IN (...) delete). This
 * turns a bulk import of N transactions into a handful of requests while
 * preserving queue ordering (we never reorder across tables, so FK ordering
 * like account-before-txn is untouched). PATCH stays per-row because two
 * updates to the same table can touch different columns.
 */

import com.powersync.PowerSyncDatabase
import com.powersync.connectors.PowerSyncBackendConnector
import com.powersync.connectors.PowerSyncCredentials
import com.powersync.db.crud.CrudEntry
import com.powersync.db.crud.UpdateType
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.postgrest.postgrest
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
// P2.3 diagnostics wiring. Unlike Classification/FailureDetails (used only
// via inferred chained member access on same-package Quarantine.kt types,
// which needs no import), these are bare top-level identifiers / a type
// named directly, from OTHER packages -- Kotlin, like Swift, requires each
// file to import what it names explicitly, regardless of what a same-package
// sibling file already imports.
import com.sanvya.app.data.diagnostics.logDiagnostic
import com.sanvya.app.domain.diagnostics.DetailValue
import com.sanvya.app.domain.diagnostics.LOG_LEVEL_ERROR
import com.sanvya.app.domain.diagnostics.LOG_LEVEL_WARN
import com.sanvya.app.domain.syncpolicy.FAILURE_CLASS_PERMANENT

/** Postgres schema that holds all Sanvya tables. */
const val DB_SCHEMA = "pocketcare"

/**
 * Optional sink for structured sync failures.
 *
 * A module-level hook rather than a constructor option, so this package stays
 * free of any dependency on the app's diagnostics layer. The app opts in by
 * calling setSyncDiagnosticSink at startup.
 */
data class SyncDiagnostic(
    val table: String,
    val op: String,
    val rows: Int,
    val code: String? = null,
    val message: String? = null,
    val hint: String? = null,
    val cls: String? = null,
    val attempts: Int? = null,
    val quarantined: Boolean? = null,
)

private var diagnosticSink: ((SyncDiagnostic) -> Unit)? = null

fun setSyncDiagnosticSink(fn: ((SyncDiagnostic) -> Unit)?) {
    diagnosticSink = fn
}

/**
 * FAULT INJECTION — dev only.
 *
 * Makes unreachable paths reachable: force uploads to a chosen table to fail
 * with a chosen SQLSTATE, and exercise the full classify → retry → quarantine
 * → recover path deliberately. Identical to connector.ts's fault injection.
 */
data class FaultInjection(
    /** Unqualified table name, or "*" for everything. */
    val table: String,
    /** SQLSTATE to report, e.g. "42501" (RLS) or "23503" (foreign key). */
    val code: String,
    val message: String? = null,
    val status: Int? = null,
)

private var fault: FaultInjection? = null

/** Install (or clear) a fault. No-op in production builds — see the app's gate. */
fun setFaultInjection(f: FaultInjection?) {
    fault = f
}

fun getFaultInjection(): FaultInjection? = fault

private fun injectedErrorFor(table: String): FailureDetails? {
    val f = fault ?: return null
    val bare = if (table.contains(".")) table.substringAfterLast(".") else table
    if (f.table != "*" && f.table != bare) return null
    return FailureDetails(
        code = f.code,
        message = f.message ?: "injected fault (${f.code}) for $bare",
        status = f.status,
    )
}

/**
 * Keys we've recorded a failure for this session (in-memory; cleared on
 * success). Only ops we know have failed can have a counter.
 */
private val bumped = mutableSetOf<String>()

private fun reportSyncDiagnostic(d: SyncDiagnostic) {
    try {
        diagnosticSink?.invoke(d)
    } catch (_: Exception) {
        // Diagnostics must never make a sync failure worse than it already is.
    }
}

/**
 * Convert a PowerSync CrudEntry to our CrudOpSummary for quarantine tracking.
 * Reads the op's data map and serialises it to JSON for dead-letter storage.
 *
 * opData.typed, NOT opData.toMap(): CrudEntry.opData's real type is
 * `SqliteRow?` (confirmed via powersync-kotlin 1.13.0's real source,
 * CrudEntry.kt/SqliteRow.kt at the pinned tag), and SqliteRow itself
 * implements `Map<String, String?>` — every value stringified, kept only
 * for backwards compatibility. `opData?.toMap()` copies that same
 * stringified view; because Kotlin's Map is declared `out V` (covariant),
 * assigning a `Map<String, String?>` into a `Map<String, Any?>`-typed val
 * compiles with ZERO warning or error, so this shipped silently wrong.
 * `.typed: Map<String, Any?>` is SqliteRow's real typed accessor (Int/
 * Double/Boolean/String) — mirrors the exact fix applied to iOS's
 * SupabaseConnector.swift (opData vs opDataTyped) after finding the same
 * class of bug there. See CAUTION below `toJsonPrimitive` for a real
 * constraint of `.typed` itself (Int, never Long, for numeric columns).
 */
private fun CrudEntry.toOpSummary(): CrudOpSummary {
    val rawData: Map<String, Any?> = opData?.typed ?: emptyMap()
    return CrudOpSummary(
        // clientId is a non-nullable Int in the real SDK (confirmed via
        // CrudEntry.kt source) — the `?.` here is a harmless-but-misleading
        // unnecessary safe call, not a bug; kept as `.toLong()` since
        // CrudOpSummary.clientId is Long?.
        clientId = clientId.toLong(),
        table = table,
        op = op.name,
        id = id,
        payloadJson = encodePayload(rawData),
    )
}

/**
 * Convert an opData value (Any?) to a kotlinx.serialization JsonElement.
 * Used to build JsonObject payloads for supabase-kt's Serializable PostgREST API.
 *
 * CAUTION (real constraint, confirmed via SqliteRow.kt source): `.typed`'s
 * numeric parser (`jsonNumberOrBoolean()`) always produces Kotlin `Int`
 * (32-bit) for non-decimal numeric strings, never `Long` — there is no
 * bigint-range path. A synced integer column whose value exceeds
 * Int32.MAX_VALUE (~2.15 billion) would silently overflow when read via
 * `.typed`. Money is stored as integer minor units (golden rule #1) and
 * unlikely to hit that range in practice, but this is an SDK-level
 * limitation, not something fixable from our side — flag for review if a
 * future schema ever needs a genuinely bigint-range synced column.
 */
private fun toJsonPrimitive(v: Any?): kotlinx.serialization.json.JsonElement = when (v) {
    null -> JsonNull
    is String -> JsonPrimitive(v)
    is Number -> JsonPrimitive(v)
    is Boolean -> JsonPrimitive(v)
    else -> JsonPrimitive(v.toString())
}

/** Extract FailureDetails from a PostgREST or network exception. */
private fun extractFailure(e: Throwable): FailureDetails = FailureDetails(
    code = (e as? io.github.jan.supabase.exceptions.RestException)?.error,
    message = e.message,
    status = (e as? io.github.jan.supabase.exceptions.HttpRequestException)?.let {
        // supabase-kt wraps HTTP errors; extract status if available
        null // status is in the cause chain — best effort
    },
)

class SupabaseConnector(
    private val client: SupabaseClient,
    private val powerSyncUrl: String,
    /** Schema the tables live in; must be an Exposed schema in Supabase settings. */
    private val schema: String = DB_SCHEMA,
) : PowerSyncBackendConnector() {

    override suspend fun fetchCredentials(): PowerSyncCredentials? {
        val session = client.auth.currentSessionOrNull() ?: return null
        return PowerSyncCredentials(
            endpoint = powerSyncUrl,
            token = session.accessToken,
        )
    }

    override suspend fun uploadData(database: PowerSyncDatabase) {
        val batch = database.getCrudBatch() ?: return

        val ops: List<CrudEntry> = batch.crud

        // Track which keys gained a bump counter so we can clear them on success.
        val bumpedThisRun = mutableSetOf<String>()

        var i = 0
        while (i < ops.size) {
            val op = ops[i]
            // 2-arg operator get(schema, table) — NOT the 1-arg get(table), which uses
            // Postgrest.Config.defaultSchema ("public" unless configured at client
            // construction, which nothing here does). Golden rule #3 (CLAUDE.md):
            // every call must be schema-qualified or PostgREST 404s. Verified against
            // supabase-kt's real Postgrest.kt source (github.com/supabase-community/
            // supabase-kt) before using this form — `operator fun get(schema: String,
            // table: String): PostgrestQueryBuilder` is a real overload, not a guess.
            val rel = client.postgrest[schema, op.table]

            // Extend the run while the next op targets the same table + op type.
            // PATCH always stays per-row (different columns may be touched).
            var j = i + 1
            while (
                j < ops.size &&
                ops[j].table == op.table &&
                ops[j].op == op.op &&
                op.op != UpdateType.PATCH
            ) {
                j++
            }
            val run = ops.subList(i, j)

            // Fault injection short-circuits before any real network call.
            val injected = injectedErrorFor("$schema.${op.table}")

            var failure: FailureDetails? = injected

            if (failure == null) {
                try {
                    when (op.op) {
                        UpdateType.PUT -> {
                            // .typed, not the bare Map<String, String?> — see
                            // the comment on toOpSummary above.
                            val rows: List<JsonObject> = run.map { entry ->
                                buildJsonObject {
                                    entry.opData?.typed?.forEach { (k, v) -> put(k, toJsonPrimitive(v)) }
                                    put("id", entry.id)
                                }
                            }
                            rel.upsert(rows)
                        }
                        UpdateType.DELETE -> {
                            val ids = run.map { it.id }
                            rel.delete { filter { isIn("id", ids) } }
                        }
                        UpdateType.PATCH -> {
                            val data: JsonObject = buildJsonObject {
                                op.opData?.typed?.forEach { (k, v) -> put(k, toJsonPrimitive(v)) }
                            }
                            rel.update(data) { filter { eq("id", op.id) } }
                        }
                    }
                } catch (e: Exception) {
                    failure = extractFailure(e)
                }
            }

            if (failure != null) {
                val summaries = run.map { it.toOpSummary() }
                val verdict = handleUploadFailure(
                    db = database,
                    run = summaries,
                    failure = failure,
                    onBump = { k ->
                        bumped.add(k)
                        bumpedThisRun.add(k)
                    },
                )

                android.util.Log.e(
                    "SanvyaSync",
                    "[Sanvya sync] upload failed for $schema.${op.table} " +
                        "(${op.op}, ${run.size} row(s), attempt ${verdict.attempts}, " +
                        "${verdict.classification.cls})" +
                        if (verdict.quarantined) " — moved to Problems syncing" else "",
                )

                // P2.3: feed the already-ported diagnostics domain (P1.6a) so a
                // future "share diagnostics" flow has a redacted, human-readable
                // record. Permanent (about to be/already quarantined) is ERROR;
                // transient (still retrying) is WARN. makeEntry redacts the raw
                // failure.message itself -- safe to pass straight through.
                logDiagnostic(
                    level = if (verdict.classification.cls == FAILURE_CLASS_PERMANENT) LOG_LEVEL_ERROR else LOG_LEVEL_WARN,
                    scope = "sync",
                    message = failure.message ?: "Upload failed for ${op.table}",
                    detail = DetailValue.Obj(
                        linkedMapOf(
                            "table" to DetailValue.Str("$schema.${op.table}"),
                            "op" to DetailValue.Str(op.op.name),
                            "rows" to DetailValue.IntNum(run.size.toLong()),
                            "attempts" to DetailValue.IntNum(verdict.attempts.toLong()),
                            "cls" to DetailValue.Str(verdict.classification.cls),
                            "quarantined" to DetailValue.Bool(verdict.quarantined),
                            "code" to (failure.code?.let { DetailValue.Str(it) } ?: DetailValue.Null),
                        ),
                    ),
                )

                reportSyncDiagnostic(
                    SyncDiagnostic(
                        table = "$schema.${op.table}",
                        op = op.op.name,
                        rows = run.size,
                        code = failure.code,
                        message = failure.message,
                        cls = verdict.classification.cls,
                        attempts = verdict.attempts,
                        quarantined = verdict.quarantined,
                    ),
                )

                if (verdict.quarantined) {
                    i = j
                    continue
                }
                throw Exception(failure.message ?: "Upload failed for ${op.table}")
            }

            // Succeeded — clear retry counters for this run so a later, unrelated
            // failure on the same row starts from zero.
            for (entry in run) {
                val key = opKey(entry.table, entry.op.name, entry.id)
                if (bumped.remove(key)) {
                    clearAttempts(database, key)
                }
            }
            i = j
        }

        // null, not "" — matches PowerSync's own official Kotlin Supabase connector
        // (github.com/powersync-ja/powersync-kotlin, integrations/supabase), which calls
        // transaction.complete(null); an empty string is a real (if likely harmless)
        // writeCheckpoint value, not "no checkpoint".
        batch.complete(null)
    }
}
